In [ ]:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


ERROR: unknown command "matplotlib.pyplot"



ModuleNotFoundError: No module named 'matplotlib'

In [4]:
import os
os.getcwd()


'c:\\Users\\PC\\PORTFOLIO\\give-me-some-credit\\notebooks'

In [6]:
os.listdir(".")



['01_eda.ipynb']

In [ ]:
os.listdir("..")



['data', 'give-me-some-credit-notebook.ipynb', 'notebooks', 'venv']

In [11]:
df= pd.read_csv("../data/cs-training.csv")
df.head(10)

,Unnamed: 0,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
0,1,1,0.766127,45,2,0.802982,9120.0,13,0,6,0,2.0
1,2,0,0.957151,40,0,0.121876,2600.0,4,0,0,0,1.0
2,3,0,0.658180,38,1,0.085113,3042.0,2,1,0,0,0.0
3,4,0,0.233810,30,0,0.036050,3300.0,5,0,0,0,0.0
4,5,0,0.907239,49,1,0.024926,63588.0,7,0,1,0,0.0
5,6,0,0.213179,74,0,0.375607,3500.0,3,0,1,0,1.0
6,7,0,0.305682,57,0,5710.000000,NaN,8,0,3,0,0.0
7,8,0,0.754464,39,0,0.209940,3500.0,8,0,0,0,0.0
8,9,0,0.116951,27,0,46.000000,NaN,2,0,0,0,NaN
9,10,0,0.189169,57,0,0.606291,23684.0,9,0,4,0,2.0


In [12]:
df.shape


(150000, 12)

In [13]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 12 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   Unnamed: 0                            150000 non-null  int64  
 1   SeriousDlqin2yrs                      150000 non-null  int64  
 2   RevolvingUtilizationOfUnsecuredLines  150000 non-null  float64
 3   age                                   150000 non-null  int64  
 4   NumberOfTime30-59DaysPastDueNotWorse  150000 non-null  int64  
 5   DebtRatio                             150000 non-null  float64
 6   MonthlyIncome                         120269 non-null  float64
 7   NumberOfOpenCreditLinesAndLoans       150000 non-null  int64  
 8   NumberOfTimes90DaysLate               150000 non-null  int64  
 9   NumberRealEstateLoansOrLines          150000 non-null  int64  
 10  NumberOfTime60-89DaysPastDueNotWorse  150000 non-null  int64  
 11  

In [14]:
df.describe()

,Unnamed: 0,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
count,150000.000000,150000.000000,150000.000000,150000.000000,150000.000000,150000.000000,1.202690e+05,150000.000000,150000.000000,150000.000000,150000.000000,146076.000000
mean,75000.500000,0.066840,6.048438,52.295207,0.421033,353.005076,6.670221e+03,8.452760,0.265973,1.018240,0.240387,0.757222
std,43301.414527,0.249746,249.755371,14.771866,4.192781,2037.818523,1.438467e+04,5.145951,4.169304,1.129771,4.155179,1.115086
min,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000
25%,37500.750000,0.000000,0.029867,41.000000,0.000000,0.175074,3.400000e+03,5.000000,0.000000,0.000000,0.000000,0.000000
50%,75000.500000,0.000000,0.154181,52.000000,0.000000,0.366508,5.400000e+03,8.000000,0.000000,1.000000,0.000000,0.000000
75%,112500.250000,0.000000,0.559046,63.000000,0.000000,0.868254,8.249000e+03,11.000000,0.000000,2.000000,0.000000,1.000000
max,150000.000000,1.000000,50708.000000,109.000000,98.000000,329664.000000,3.008750e+06,58.000000,98.000000,54.000000,98.000000,20.000000


Vemos que las columnas RevolvingUtilizationOfUnsecuredLines DebtRatio	MonthlyIncome	presentan outliers y la columna MonthlyIncome y numberofdependents presentan valores nulos o faltantes.

In [15]:
df.duplicated()


0         False
1         False
2         False
3         False
4         False
          ...  
149995    False
149996    False
149997    False
149998    False
149999    False
Length: 150000, dtype: bool

Vemos que no hay duplicados

Limpiando valores de la columna RevolvingUtilizationOfUnsecuredLines

In [17]:
df[df['RevolvingUtilizationOfUnsecuredLines']>1.0]


,Unnamed: 0,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
162,163,1,1.046279,47,1,1.104301,5416.0,6,0,2,0,1.0
191,192,0,1.095083,53,5,0.536704,3500.0,7,1,1,2,0.0
226,227,1,1.953488,38,0,0.153500,3556.0,5,0,0,0,2.0
251,252,1,1.048211,58,7,0.151957,3500.0,12,0,0,0,0.0
293,294,0,2340.000000,45,0,0.339333,8333.0,7,0,2,0,2.0
...,...,...,...,...,...,...,...,...,...,...,...,...
149939,149940,0,1.049900,26,0,0.073220,6500.0,6,0,0,1,0.0
149955,149956,1,1.135552,41,2,0.845887,7500.0,12,0,4,1,0.0
149962,149963,0,1.005733,48,0,0.248896,2944.0,4,3,0,0,1.0
149964,149965,0,1.010934,63,0,0.608211,6015.0,8,0,1,0,0.0


In [18]:
#Número de datos desviados
len(df[df['RevolvingUtilizationOfUnsecuredLines'] > 1.0])

3321

In [19]:
#Porcentaje de los datos desviados
len(df[df['RevolvingUtilizationOfUnsecuredLines'] > 1.0]) / len(df) * 100

2.214

In [20]:
# Ver los 5 valores más grandes
df[df['RevolvingUtilizationOfUnsecuredLines'] > 1.0]['RevolvingUtilizationOfUnsecuredLines'].nlargest(5)


85489     50708.0
31414     29110.0
16956     22198.0
149160    22000.0
149279    20514.0
Name: RevolvingUtilizationOfUnsecuredLines, dtype: float64

In [22]:
# Ver todos los valores > 1.0
outliers = df[df['RevolvingUtilizationOfUnsecuredLines'] > 1.0]['RevolvingUtilizationOfUnsecuredLines']



In [23]:
# Ver cuántos están en cada rango
print("Valores entre 1.0 y 1.1:")
print(len(outliers[outliers <= 1.1]))

print("\nValores entre 1.1 y 2.0:")
print(len(outliers[(outliers > 1.1) & (outliers <= 2.0)]))

print("\nValores entre 2.0 y 10.0:")
print(len(outliers[(outliers > 2.0) & (outliers <= 10.0)]))

print("\nValores > 10.0:")
print(len(outliers[outliers > 10.0]))

# Ver algunos ejemplos de cada rango
print("\n--- Ejemplos entre 1.0 y 1.1 ---")
print(outliers[outliers <= 1.1].head(10).values)

print("\n--- Ejemplos entre 1.1 y 2.0 ---")
print(outliers[(outliers > 1.1) & (outliers <= 2.0)].head(10).values)

print("\n--- Ejemplos > 10.0 ---")
print(outliers[outliers > 10.0].head(10).values)



Valores entre 1.0 y 1.1:
1869

Valores entre 1.1 y 2.0:
1081

Valores entre 2.0 y 10.0:
130

Valores > 10.0:
241

--- Ejemplos entre 1.0 y 1.1 ---
[1.0462791  1.09508336 1.04821084 1.00264667 1.04749591 1.01133279
 1.00850653 1.00442167 1.0085156  1.08982585]

--- Ejemplos entre 1.1 y 2.0 ---
[1.95348837 1.24584718 1.24875125 1.16318411 1.25748503 1.36214175
 1.70804492 1.62446507 1.20930233 1.63455149]

--- Ejemplos > 10.0 ---
[2340. 2066. 1143. 6324. 1982. 3746.  941. 8497. 1577. 1028.]


Aca se entiende que los valores entre 1.0 y 1.1 son valores que estan en el limite y no se reemplazan
Por otra parte los valores entre 1.1 y 10.0 son posibles errores de escala por lo que los dividimos por 100
Por último los valores mayores 10.0 son errores graves que seran reemplazados por 1.0

In [24]:
# Valores entre 1.1 y 10.0: Posibles errores de escala → dividir por 100
df.loc[(df['RevolvingUtilizationOfUnsecuredLines'] > 1.1) & 
       (df['RevolvingUtilizationOfUnsecuredLines'] <= 10.0), 
       'RevolvingUtilizationOfUnsecuredLines'] = df['RevolvingUtilizationOfUnsecuredLines'] / 100

# Valores > 10.0: Errores graves → reemplazar con 1.0
df.loc[df['RevolvingUtilizationOfUnsecuredLines'] > 10.0, 
       'RevolvingUtilizationOfUnsecuredLines'] = 1.0

In [30]:
df['RevolvingUtilizationOfUnsecuredLines'].describe()

count    150000.000000
mean          0.311647
std           0.345909
min           0.000000
25%           0.028097
50%           0.146969
75%           0.539586
max           1.099933
Name: RevolvingUtilizationOfUnsecuredLines, dtype: float64

Limpiando valores de la columna DebtRatio

In [34]:
print("Valores > 5.0:")
print(len(df[df['DebtRatio'] > 5.0]))

print("\nValores > 10.0:")
print(len(df[df['DebtRatio'] > 10.0]))

print("\nValor máximo:")
print(df['DebtRatio'].max())

print("\nPercentil 99:")
print(df['DebtRatio'].quantile(0.99))

Valores > 5.0:
29646

Valores > 10.0:
28877

Valor máximo:
329664.0

Percentil 99:
4979.040000000037


Observamos que el número de valores por encima de 5.0 y de 10.0 es muy alto y que el valor máximo tambien por lo que veremos si el problema esta en esos 5% de percentil o más

In [38]:
print("Percentil 95:")
print(df['DebtRatio'].quantile(0.95))

Percentil 95:
0.9823854748999992


En este caso habiamos observado que no eran solo el 5% por lo que podemos deducir que uno de los problemas por el cual el debratio daba tan alto es que el monthlyincome era de 0 entonces eso potenciaba todos los valores, por lo que decidimos eliminar el monthlyincome = 0 como lo vemos a continuación

In [36]:
print("Valores de MonthlyIncome = 0:")
print(len(df[df['MonthlyIncome'] == 0]))

print("\nValores de MonthlyIncome entre 0 y 500:")
print(len(df[(df['MonthlyIncome'] > 0) & (df['MonthlyIncome'] < 500)]))

Valores de MonthlyIncome = 0:
1634

Valores de MonthlyIncome entre 0 y 500:
1059


In [37]:
# Paso 1: Eliminar MonthlyIncome = 0
df = df[df['MonthlyIncome'] > 0]

In [39]:
print("DebtRatio describe:")
print(df['DebtRatio'].describe())

print("\nValores > 5.0:")
print(len(df[df['DebtRatio'] > 5.0]))

DebtRatio describe:
count    118635.000000
mean          5.291856
std         199.774225
min           0.000000
25%           0.141496
50%           0.292454
75%           0.473289
max       61106.500000
Name: DebtRatio, dtype: float64

Valores > 5.0:
862


aca observamos que los valores con 5.0 por encima de la media son realmente muy pocos, 862 (0,58%) por lo que, para mantener la integridad de los datos decidimos deshacernos de esas columnas, ya que representan muy pocas cantidades.

In [41]:
df = df[df['DebtRatio'] <= 5.0]
print(df['DebtRatio'].describe())

count    117773.000000
mean          0.359841
std           0.369973
min           0.000000
25%           0.140384
50%           0.290280
75%           0.467914
max           5.000000
Name: DebtRatio, dtype: float64


In [42]:
print("DebtRatio limpio:")
print(df['DebtRatio'].describe())

print("\nValores > 5.0:")
print(len(df[df['DebtRatio'] > 5.0]))

print("\nShape actual:")
print(df.shape)

DebtRatio limpio:
count    117773.000000
mean          0.359841
std           0.369973
min           0.000000
25%           0.140384
50%           0.290280
75%           0.467914
max           5.000000
Name: DebtRatio, dtype: float64

Valores > 5.0:
0

Shape actual:
(117773, 12)


Aca terminamos de limpiar los valores de DebtRatio

Limpiando valores de la columna MonthlyIncome

In [44]:
print("MonthlyIncome NaN:")
print(df['MonthlyIncome'].isna().sum())

print("\nMonthlyIncome describe:")
print(df['MonthlyIncome'].describe())

print("\nValores > 500,000:")
print(len(df[df['MonthlyIncome'] > 500000]))

print("\nPercentil 99:")
print(df['MonthlyIncome'].quantile(0.99))
print("\nPercentil 95:")
print(df['MonthlyIncome'].quantile(0.95))

MonthlyIncome NaN:
0

MonthlyIncome describe:
count    1.177730e+05
mean     6.805639e+03
std      1.450415e+04
min      1.000000e+00
25%      3.500000e+03
50%      5.494000e+03
75%      8.333000e+03
max      3.008750e+06
Name: MonthlyIncome, dtype: float64

Valores > 500,000:
12

Percentil 99:
25000.0

Percentil 95:
14719.599999999977


In [45]:
df = df[df['MonthlyIncome'] <= 500000]

print("Valores > 500k después:")
print(len(df[df['MonthlyIncome'] > 500000]))

print("\nMonthlyIncome describe:")
print(df['MonthlyIncome'].describe())

print("\nShape final:")
print(df.shape)

Valores > 500k después:
0

MonthlyIncome describe:
count    117761.000000
mean       6697.413405
std        6775.529703
min           1.000000
25%        3500.000000
50%        5492.000000
75%        8333.000000
max      440000.000000
Name: MonthlyIncome, dtype: float64

Shape final:
(117761, 12)


In [46]:
print("=== VERIFICACIÓN FINAL ===\n")

print("Shape:", df.shape)

print("\nNaN por columna:")
print(df.isna().sum())

print("\n=== DESCRIBE FINAL ===")
print(df.describe())

=== VERIFICACIÓN FINAL ===

Shape: (117761, 12)

NaN por columna:
Unnamed: 0                              0
SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64

=== DESCRIBE FINAL ===
          Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  \
count  117761.000000     117761.000000                         117761.000000   
mean    75032.251526          0.070066                              0.323027   
std     43280.494111          0.255259                              0.343824   
min         1.000000          0.000000                              0.000

In [ ]:
df = df.drop('Unnamed: 0', axis=1)


In [49]:
print(df.columns)

Index(['SeriousDlqin2yrs', 'RevolvingUtilizationOfUnsecuredLines', 'age',
       'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio', 'MonthlyIncome',
       'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate',
       'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse',
       'NumberOfDependents'],
      dtype='object')
